### DISCUSSION TITLE
How to override `injections.scm`?

### DISCUSSION THREAD

**Discussion by rafikdraoui** [nvim-treesitter/nvim-treesitter#958] [Q&A] [answered]
I'm trying to disable `@rst` injections in Python docstrings.

At first, I thought I could create my own `queries/python/injections.scm` to override nvim-treesitter's, but that doesn't work. (c.f. https://github.com/nvim-treesitter/nvim-treesitter/tree/5757f8a50d5d26b8c184b3a51713db763cdd9702#adding-queries)

Another idea is to create `after/queries/python/injections.scm`, and try to "cancel" the docstrings injection rules, but it's not clear to me how to do this (replacing all the `@rst` by `@string` or `@none` doesn't seem to work), and in any case, this feels more fragile than just overriding the whole file.

https://github.com/nvim-treesitter/nvim-treesitter/issues/839 mentions that overriding queries is an upstream issue, so maybe that's what is happening here. If that's the case, it might be worth putting a note in the "Adding queries" section of the README about it, so that people don't get surprised when things work differently than what's described in the docs :smiley: 

As an aside, the documentation mentions that the `use_languagetree` setting in the `highlight` config can be used to disable injections, but the corresponding code checking for that config value has been removed (in any case, I wouldn't want to disable injections altogether). https://github.com/nvim-treesitter/nvim-treesitter/blob/5757f8a50d5d26b8c184b3a51713db763cdd9702/doc/nvim-treesitter.txt#L113

**Comment by theHamsta** [comment]
@steelsojka @bfredl with the new "just rely on rtp" it's no longer easy to replace a runtime file that a plugin dictates by placing your version of this file before it. What would you suggest users that don't like a specific query file of a plugin?

I personally don't like that some parts of do strings are highlighted white. Since most python doc strings are not really rst, the highlighting is not really accurate and just  some random parts of it get different highlights (eg the first line is always bold but only if it's on the same line as the """)

**Comment by rafikdraoui** [comment, marked as answer]
It is now possible to override queries with the `vim.treesitter.set_query` function, so there is a workaround for this: https://github.com/nvim-treesitter/nvim-treesitter/pull/1204#issuecomment-873416709

https://github.com/nvim-treesitter/nvim-treesitter/pull/1240 might provide a better way to do this in the future.

**Comment by lf-** [comment]
(I know that this is a necropost but someone might find this useful in the future)

You can also do overrides by monkeypatching the query loading machinery to allow disabling the extension mechanism and load only the first file, which is in `~/.config/nvim/queries/haskell/textobjects.scm`:

```lua
-- every query in here will use only the first file, thereby overriding the
-- query retrieved to ignore ones from plugins, since ~/.config/nvim is earlier
-- in the runtime path
local OVERRIDE_QUERIES = {
    haskell = {
        textobjects = true,
    }
}

-- monkeypatch get_query_files to allow for queries to be overridden.
-- this needs to be very early in initialization so that nvim-treesitter hits it

-- stop require'vim.treesitter' from returning a different module than vim
-- already loaded (!)
package.loaded['vim.treesitter'] = vim.treesitter
package.loaded['vim.treesitter.query'] = vim.treesitter.query
local _get_query_files = vim.treesitter.get_query_files

vim.treesitter.get_query_files = function(lang, query_name, is_included)
    local files = _get_query_files(lang, query_name, is_included)

    if OVERRIDE_QUERIES[lang] and OVERRIDE_QUERIES[lang][query_name] then
        files = vim.list_slice(files, 1, 1)
    end
    return files
end
-- lol lmao 😭
vim.treesitter.query.get_query_files = vim.treesitter.get_query_files
```
